File: day05.py
from hashlib import md5

door_id = ''
with open('input/05') as f:
    door_id = f.read().rstrip('\n')

password = ''
i = 0
while len(password) < 8:
    hexed = md5((door_id + str(i)).encode('utf-8')).hexdigest()
    if hexed.startswith('00000'):
        password += hexed[5]
    i += 1

print('Password:', password)

password_p2 = [None] * 8
i = 0
while None in password_p2:
    hexed = md5((door_id + str(i)).encode('utf-8')).hexdigest()
    if hexed.startswith('00000'):
        try:
            position = int(hexed[5])
        except ValueError:
            i += 1
            continue
        char = hexed[6]
        if position < len(password_p2) and password_p2[position] is None:
            password_p2[position] = char
    i += 1
print('Password (Part2):', ''.join(password_p2))

File: day05.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func is_full(s []string) bool {
	for _, a := range s {
		if a == "" {
			return false
		}
	}
	return true
}

func main() {
	file, err := os.Open("input/05")
	if err != nil {
		log.Fatal(err)
	}
	rawDoorID := make([]byte, 100)
	_, err = file.Read(rawDoorID)
	if err != nil {
		log.Fatal(err)
	}
	doorID := strings.Split(string(rawDoorID), "\n")[0]
	// PART 1
	password := ""
	for i := 0; len(password) < 8; i++ {
		// build string
		merged := doorID + strconv.Itoa(i)
		// md5
		hashed := md5.Sum([]byte(merged))
		// hex
		hexed := hex.EncodeToString(hashed[:])
		// check for leading zeros
		if strings.HasPrefix(hexed, "00000") {
			password += string(hexed[5])
		}
	}
	fmt.Printf("Password (Part1): %v\n", password)
	// PART 2
	password2 := make([]string, 8)
	for i := 0; !is_full(password2); i++ {
		// build string
		merged := doorID + strconv.Itoa(i)
		// md5
		hashed := md5.Sum([]byte(merged))
		// hex
		hexed := hex.EncodeToString(hashed[:])
		// check for leading zeros
		if strings.HasPrefix(hexed, "00000") {
			position, err := strconv.Atoi(string(hexed[5]))
			if err != nil {
				continue
			}
			if position < len(password2) && password2[position] == "" {
				password2[position] = string(hexed[6])
			}
		}
	}
	fmt.Printf("Password (Part2): %v\n", strings.Join(password2, ""))
}
